server/protocols/protocols-smtp/src/test/java/org/apache/james/smtpserver/ValidRcptHandlerTest.java

```diff
@@ -103,63 +111,96 @@ public Object getAttachment(String key, State state) {
+        SMTPSession session = setupMockedSMTPSession(RELAYING_ALLOWED);
 
-        int rCode = handler.doRcpt(session, null, mailAddress).getResult();
+        int rCode = handler.doRcpt(session, SENDER, invalidUserEmail).getResult();
 
         assertThat(rCode).isEqualTo(HookReturnCode.DENY);
     }
 
     @Test
-    public void testRejectInvalidUserRelay() throws Exception {
-        MailAddress mailAddress = new MailAddress(INVALID_USER + "@localhost");
-        SMTPSession session = setupMockedSMTPSession(true);
+    public void doRcptShouldDeclineNonLocalUsersWhenRelay() throws Exception {
+        MailAddress mailAddress = new MailAddress(INVALID_USER + "@otherdomain");
+        SMTPSession session = setupMockedSMTPSession(RELAYING_ALLOWED);
 
-        int rCode = handler.doRcpt(session, null, mailAddress).getResult();
+        int rCode = handler.doRcpt(session, SENDER, mailAddress).getResult();
 
-        assertThat(rCode).isEqualTo(HookReturnCode.DENY);
+        assertThat(rCode).isEqualTo(HookReturnCode.DECLINED);
+    }
+    
+    @Test
+    public void doRcptShouldDeclineNonLocalUsersWhenNoRelay() throws Exception {
+        MailAddress mailAddress = new MailAddress(INVALID_USER + "@otherdomain");
+        SMTPSession session = setupMockedSMTPSession(!RELAYING_ALLOWED);
+
+        int rCode = handler.doRcpt(session, SENDER, mailAddress).getResult();
+
+        assertThat(rCode).isEqualTo(HookReturnCode.DECLINED                         );
     }
 
     @Test
-    public void testNotRejectValidUser() throws Exception {
-        MailAddress mailAddress = new MailAddress(VALID_USER + "@localhost");
-        SMTPSession session = setupMockedSMTPSession(false);
+    public void doRcptShouldDeclineValidUsersWhenNoRelay() throws Exception {
+        SMTPSession session = setupMockedSMTPSession(!RELAYING_ALLOWED);
 
-        int rCode = handler.doRcpt(session, null, mailAddress).getResult();
+        int rCode = handler.doRcpt(session, SENDER, validUserEmail).getResult();
 
         assertThat(rCode).isEqualTo(HookReturnCode.DECLINED);
     }
 
     @Test
-    public void testHasAddressMapping() throws Exception {
-        MailAddress mailAddress = new MailAddress(USER1 + "@localhost");
-        SMTPSession session = setupMockedSMTPSession(false);
+    public void doRcptShouldDeclineValidUsersWhenRelay() throws Exception {
+        SMTPSession session = setupMockedSMTPSession(RELAYING_ALLOWED);
 
-        int rCode = handler.doRcpt(session, null, mailAddress).getResult();
+        int rCode = handler.doRcpt(session, SENDER, validUserEmail).getResult();
 
         assertThat(rCode).isEqualTo(HookReturnCode.DECLINED);
     }
 
     @Test
-    public void testHasErrorMapping() throws Exception {
-        MailAddress mailAddress = new MailAddress(USER2 + "@localhost");
-        SMTPSession session = setupMockedSMTPSession(false);
+    public void doRcptShouldDeclineWhenHasAddressMapping() throws Exception {
```

Comment:
why?

Reply:
Because it do mean we care about that address as it is being redirected...